### PR TITLE
Table of thematics per phase devas 1639

### PR DESCRIPTION
**Pull request by amensouissi**

### CODE REVIEW

File: assembl/static2/css/common.scss [hunk 1]
```diff
@@ -271,6 +271,87 @@ input.placeholder, textarea.placeholder {
...
+      }
+    }
+
+    .form-tree-item::after {
+      position: absolute;
+      content: "";
+      top: 18px;
+      left: -15px;
+      border-left: 1px solid #ccc;
+      border-top: 1px solid #ccc;
```
Comment (paolina92) [inline assembl/static2/css/common.scss]: Use an existing variable in variables.scss

File: assembl/static2/css/common.scss [hunk 2]
```diff
@@ -271,6 +271,87 @@ input.placeholder, textarea.placeholder {
...
+      top: 8px;
+    }
+
+    .form-tree-item::before {
+      content: "";
+      position: absolute;
+      top: -26px;
+      left: -15px;
+      border-left: 1px solid #ccc;
+      border-bottom: 1px solid #ccc;
```
Comment (paolina92) [inline assembl/static2/css/common.scss]: Same here

**Comment by vincentfretin** [comment]
Your new code should be used for thread and brightMirror phases as well, not only survey.

File: assembl/static2/js/app/graphql/ThematicsDataQuery.graphql [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+      id
+      title(lang: $lang)
+    }
+    ... on Idea {
+      id
+      parentId
+      title(lang: $lang)
+    }
+  }
+  rootIdea {
```
Comment (vincentfretin) [inline assembl/static2/js/app/graphql/ThematicsDataQuery.graphql]: very important, `rootIdea((identifier: $identifier)` (discussionPhaseId in the future)

File: assembl/static2/js/app/graphql/ThematicsQuery.graphql [hunk 4]
```diff
@@ -58,4 +51,12 @@ query ThematicsQuery($identifier: String!) {
       }
     }
   }
+  rootIdea {
```
Comment (vincentfretin) [inline assembl/static2/js/app/graphql/ThematicsQuery.graphql]: `rootIdea(identifier: $identifier)` (discussionPhaseId in the future)

File: assembl/static2/js/app/utils/translations.js [hunk 5]
```diff
@@ -550,9 +552,13 @@ const Translations = {
...
       survey: {
-        "0": "Renseigner les thématiques",
-        "1": "Renseigner les questions",
-        "2": "Exporter les données"
+        "1": "Créer la table des thématiques",
+        "2": "Renseigner les questions",
+        "3": "Exporter les données",
+        configThematic: "Configurer la thématique",
+        configThematics: "Configurer les thématiques",
+        configThematicsHlperTitle: "Vous avez choisie de configurer une table de thématiques.",
```
Comment (cedricmessiant) [inline assembl/static2/js/app/utils/translations.js]: "Vous avez choisi"

File: assembl/static2/js/app/utils/tree.js [hunk 6]
```diff
@@ -27,11 +32,93 @@ export function getChildren<T: NodeType>(rootNode: T, nodes: Array<T>): Array<T>
...
+      let children = nodes.filter(n => n.parentId === ancestor.id);
+      children = sortItems ? sortItems(children) : children;
+      path.push(children.indexOf(node));
+      node = ancestor;
+    });
+    if (roots.includes(node)) {
+      path.push(sortItems(roots).indexOf(node));
+    }
+  }
+  return path.reverse();
```
Comment (cedricmessiant) [inline assembl/static2/js/app/utils/tree.js]: I think you talked about it during stand-up, we could use unit tests for all these functions

File: assembl/static2/js/app/components/administration/survey/configureThematicForm.jsx [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        editLocale={editLocale}
+        name={`${name}.title`}
+        component={MultilingualTextFieldAdapter}
+        label={`${I18n.t('administration.tableOfThematics.thematicTitle')} ${editLocale.toUpperCase()}`}
+      />
+      <Field
+        deleteTooltip={deleteThematicImageTooltip}
+        name={`${name}.img`}
+        component={FileUploaderFieldAdapter}
+        label={I18n.t('administration.voteSessionHeaderLabel')}
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/static2/js/app/components/administration/survey/configureThematicForm.jsx]: old code ;)

File: assembl/static2/js/app/components/administration/survey/index.jsx [hunk 8]
```diff
@@ -40,15 +44,18 @@ const DumbSurveyAdminForm = ({ client, currentStep, debateId, editLocale, locale
       <React.Fragment>
         <div className="admin-content">
           <AdminForm handleSubmit={handleSubmit} pristine={pristine} submitting={submitting}>
-            {currentStep === 1 && <Step1 editLocale={editLocale} />}
-            {currentStep === 2 && <Step2 editLocale={editLocale} values={values} />}
-            {currentStep === 3 && <Step3 debateId={debateId} locale={locale} />}
+            {section === '1' && <Step1 editLocale={editLocale} locale={locale} />}
+            {section === 'configThematics' && (
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/static2/js/app/components/administration/survey/index.jsx]: it's the case

File: assembl/static2/js/app/utils/tree.js (hunk 6 above)
Comment (amensouissi) [reply assembl/static2/js/app/utils/tree.js]: yes

File: assembl/static2/js/app/utils/tree.js [hunk 9]
```diff
@@ -27,11 +32,91 @@ export function getChildren<T: NodeType>(rootNode: T, nodes: Array<T>): Array<T>
  */
 export function getPartialTree<T: NodeType>(nodes: Array<T>): TreeType<T> {
   let ids = nodes.map(node => node.id);
-  const roots = nodes.filter(node => node.ancestors.every(a => !ids.includes(a)));
+  const roots = nodes.filter(node => !node.ancestors || node.ancestors.every(a => !ids.includes(a)));
```
Comment (cedricmessiant) [inline assembl/static2/js/app/utils/tree.js]: I don't understand how a tree can have multiple roots (same for `getPartialTreeByParent` below). Maybe a comment would be welcome here to clarify this

Comment (amensouissi) [reply assembl/static2/js/app/utils/tree.js]: not a tree but a Partial Tree ;)

File: assembl/static2/js/app/components/form/fieldArrayWithActions.jsx [hunk 10]
```diff
@@ -2,81 +2,222 @@
...
-        {fields.map((fieldname, idx) => (
-          <div className="form-container" key={fieldname}>
-            {titleMsgId && (
-              <div className="title left">
-                <Translate value={titleMsgId} index={idx + 1} />
-              </div>
-            )}
+type FieldsProps = {
+  fields: FieldArrayRenderProps,
+  onAdd?: (id: string, parentId: string, index: number) => void,
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/components/form/fieldArrayWithActions.jsx]: Why did you add handlers for these actions? Why not hooks for these operations? Eg. `onBeforeAdd`, or `onAfterAdd`, etc. This way, the core of the action itself is controlled by your code, which is safer.

File: assembl/static2/js/app/graphql/mutations/deleteAcceptedCookies.graphql [hunk 11]
```diff
@@ -1,7 +0,0 @@
-mutation deleteAcceptedCookies($action: String!) {
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/graphql/mutations/deleteAcceptedCookies.graphql]: Ahh this wasn't removed in a previous PR 😱 !? Thanks for doing it here then.

File: assembl/static2/js/app/pages/multiColumnsAdmin.jsx [hunk 12]
```diff
@@ -54,7 +53,7 @@ class MultiColumnsAdmin extends React.Component {
             annotation="multicolumnAnnotation"
           />
         )}
-        {!isNaN(currentStep) && <Navbar currentStep={currentStep} totalSteps={1} phaseIdentifier="multiColumn" />}
+        {section && <Navbar currentStep={section} steps={['1']} phaseIdentifier="multiColumn" />}
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/pages/multiColumnsAdmin.jsx]: The `phaseIdentifier` should now use the one from constants, no?

File: assembl/static2/js/app/pages/surveyAdmin.jsx [hunk 13]
```diff
@@ -4,19 +4,45 @@ import { I18n } from 'react-redux-i18n';
...
 
-const SurveyAdmin = ({ section }: Props) => {
-  const currentStep = parseInt(section, 10);
-  const sectionTitleMsgId = `administration.survey.${currentStep - 1}`;
+const getSectionTitle = (section: string, thematicId: string): string | null => {
+  if (PHASES_ADMIN_MENU.survey.subMenu.configThematics && thematicId) {
+    return 'administration.survey.configThematic';
+  }
+  const menuItem = getAdminMenuSection(section, PHASES_ADMIN_MENU.survey.subMenu);
+  return (menuItem && menuItem.title) || null;
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/pages/surveyAdmin.jsx]: This could return `boolean | null` at runtime, so it's not consistent with the type signature of the method.

File: assembl/static2/js/app/utils/translations.js [hunk 14]
```diff
@@ -885,7 +897,8 @@ const Translations = {
       privacyPolicy: "Privacy policy",
       matomoSettings: "To modify this cookie's setting please click here",
       userSessionHelper: "This cookie is necessary to maintain the user connected on Assembl.",
-      localeHelper: "This cookie is necessary to display the texts in the language of your browser or the one you selected in the navigation bar.",
+      localeHelper:
+        "This cookie is necessary to display the texts in the language of your browser or the one you selected in the navigation bar.",
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/utils/translations.js]: `This cookie is necessary in order to display the page in the language of the browser or the language selected in the navigation bar.`

File: assembl/static2/js/app/utils/translations.js [hunk 15]
```diff
@@ -1318,10 +1331,11 @@ const Translations = {
       confirmTextFieldDeletion: "Are you sure that you want to delete this field?",
       confirmSelectFieldOptionDeletionTitle: "Delete this item",
       confirmSelectFieldOptionDeletion: "Are you sure that you want to delete this item?",
-      addThematic: "Add a theme",
+      addThematic: "Add a level %{level} theme",
       addQuestion: "Add a question",
       anErrorOccured: "An error occured during save. Please check that you filled all the required fields.",
       deleteThematic: "Delete the theme",
+      deleteSubThematicDisabled: "You can't delete a theme with sub-thematics.",
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/utils/translations.js]: cannot* (can't is informal)

File: assembl/static2/js/app/routes.jsx [hunk 16]
```diff
@@ -84,12 +84,12 @@ const AdminChild = (props) => {
...
     return <DiscussionAdmin {...props} section={props.location.query.section} />;
   case 'survey':
-    return <SurveyAdmin {...props} thematicId={props.location.query.thematic} section={props.location.query.section} />;
+    return <SurveyAdmin {...props} thematicId={props.location.query.thematicId} section={props.location.query.section} />;
   case 'brightMirror':
     return <BrightMirrorAdmin {...props} />;
   case 'thread':
     return <ThreadAdmin {...props} section={props.location.query.section} />;
-  case 'multiColumn':
+  case 'multiColumns':
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/routes.jsx]: Not sure if it's necessary as it's admin page, but want to swap these switch statement strings with the constants value instead?

File: assembl/static2/js/app/pages/multiColumnsAdmin.jsx (hunk 12 above)
Comment (vincentfretin) [reply assembl/static2/js/app/pages/multiColumnsAdmin.jsx]: yeah, and should probably be multiColumns with a S

File: assembl/static2/css/variables.scss [hunk 17]
```diff
@@ -11,6 +11,7 @@ $grey7: #e5e5e5;
 $grey8: #8b8b8b;
 $grey9: #efefef;
 $grey10: #f7f8fc;
+$grey11: #ccc;
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/static2/css/variables.scss]: it's the input border color (a bootstrap form-control class ).

File: assembl/static2/flow/graphql_types.flow.js [hunk 18]
```diff
@@ -13,6 +13,18 @@ export type PublicationStates =
   | 'SUBMITTED_IN_EDIT_GRACE_PERIOD'
   | 'WIDGET_SCOPED';
 
+export type CookieTypes =
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/static2/flow/graphql_types.flow.js]: flow/README ;)

File: assembl/static2/js/app/components/administration/menu.jsx [hunk 19]
```diff
@@ -1,130 +1,70 @@
...
+    const { requestedPhase } = this.props;
+    const { title, sectionId, subMenu } = menuItem;
+    const sectionIndex = rootSection ? `${rootSection}.${sectionId}` : sectionId;
+    const sectionQuery = sectionId ? `?section=${sectionIndex}` : '';
+    const subMenuIds = subMenu ? Object.keys(subMenu) : [];
+    const newRootSection = !isRoot ? sectionIndex : '';
+    const isActive = requestedPhase === id;
+    return (
+      <li key={id + sectionIndex} className={isRoot ? 'menu-item' : ''}>
+        <Link to={`${get('administration', slug)}/${id}${sectionQuery}`} activeClassName="active">
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/static2/js/app/components/administration/menu.jsx]: We need "administration": "${slug}/administration/${id}". For the query, we need to use this (which I added in the use_discussionPhaseId_DEVAS-1750 branch): https://github.com/assembl/assembl/blob/9ccdcccdc61958e3a649431f677bcadae142b8c0/assembl/static2/js/app/utils/routeMap.js#L31

File: assembl/static2/js/app/components/administration/navbar.jsx [hunk 20]
```diff
@@ -7,49 +8,62 @@ import { get } from '../../utils/routeMap';
...
   }
 
-  goToSection(stepNb) {
+  goToSection = (stepNb: number) => {
     const slug = { slug: getDiscussionSlug() };
+    const stepId = this.state.steps[stepNb - 1];
     const { phaseIdentifier } = this.props;
     browserHistory.push(
-      `${get('administration', slug)}${get('adminPhase', { ...slug, phase: phaseIdentifier })}?section=${stepNb}`
+      `${get('administration', slug)}${get('adminPhase', { ...slug, phase: phaseIdentifier })}?section=${stepId}`
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/static2/js/app/components/administration/navbar.jsx]: "administration": "${slug}/administration/${id}"

File: assembl/static2/js/app/components/administration/thematicsMenu.jsx [hunk 21]
```diff
@@ -0,0 +1,267 @@
...
+  data: { rootIdea, thematicsData },
+  location
+}: ThematicsMenuProps & Props) => {
+  if (!thematicsData) return null;
+  const sectionIndex = rootSectionId ? `${rootSectionId}.${menuItem.sectionId}` : menuItem.sectionId;
+  const sectionQuery = `?section=${sectionIndex}`;
+  const { roots, descendants } = getPartialTreeByParentId(rootIdea && rootIdea.id, thematicsData);
+  if (roots.length === 0) return null;
+  const firstThematics = sortBy(roots, 'order')[0];
+  const to = `${get('administration', slug)}${get('adminPhase', {
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/static2/js/app/components/administration/thematicsMenu.jsx]: we need to use this (which I added in the use_discussionPhaseId_DEVAS-1750 branch): https://github.com/assembl/assembl/blob/9ccdcccdc61958e3a649431f677bcadae142b8c0/assembl/static2/js/app/utils/routeMap.js#L31

File: assembl/static2/js/app/pages/surveyAdmin.jsx (hunk 13 above)
Comment (amensouissi) [reply assembl/static2/js/app/pages/surveyAdmin.jsx]: (menuItem && menuItem.title) || null

1) menuItem = {title: 'foo'} => return 'foo'
2) menuItem = null or undefined => return null || null => return null

====> string | null

File: assembl/static2/js/app/components/administration/menu.jsx (hunk 19 above)
[the remark this answers is not in the document]
Comment (ayyazdaniaryan) [reply assembl/static2/js/app/components/administration/menu.jsx]: Perfect, that would work well too. Keep the root path, query strings to be added after in the fetcher. And good job on the urlencoding of query string

File: assembl/static2/js/app/pages/surveyAdmin.jsx (hunk 13 above)
Comment (ayyazdaniaryan) [reply assembl/static2/js/app/pages/surveyAdmin.jsx]: Huh, been python-ing too long. I read `(menuItem && menuItem.title)` as returning a boolean. Fair enough.

File: assembl/static2/js/app/components/administration/thematicsMenu.jsx (hunk 21 above)
[the remark this answers is not in the document]
Comment (ayyazdaniaryan) [reply assembl/static2/js/app/components/administration/thematicsMenu.jsx]: Yeah, agreed. So we'll merge when that branch is merged after. Perhaps a TODO comment here then, thanks

File: assembl/static2/js/app/components/administration/survey/configureThematicForm.jsx [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+  editLocale: string,
+  thematicId: string,
+  values: ?SurveyAdminValues
+};
+
+export function getFieldName(themeId: string, values: ThemesValue, fieldName: string): string {
+  let result = '';
+  values.some((value, index) => {
+    if (value.id === themeId) {
+      result = `${fieldName}[${index}]`;
```
[the remark this answers is not in the document]
Comment (amensouissi) [reply assembl/static2/js/app/components/administration/survey/configureThematicForm.jsx]: we need to use a while :)
